pennylane/vqe/vqe.py

```diff
@@ -120,6 +120,39 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    def is_diagonal(self):
+        r""" Checks if a Hamiltonian is diagonal in the computational basis.
+
+        Returns:
+            bool: ``True`` if the Hamiltonian is diagonal in the computational basis, ``False`` otherwise.
+        """
+
+        non_diagonal_coeffs = []
+        non_diagonal_obs = []
+
+        # Loops through each term of the Hamiltonian
+        for i, term in enumerate(self._ops):
+
+            # Prunes identity from all tensor products
+            term = term.prune() if isinstance(term, qml.operation.Tensor) else term
+
+            # Combines like terms (provided they are non-diagonal)
+            if bool(np.count_nonzero(term.matrix - np.diag(np.diagonal(term.matrix)))):
+
+                t = [term.matrix.tolist(), term.wires]
+
+                if t in non_diagonal_obs:
+                    non_diagonal_coeffs[non_diagonal_obs.index(t)] += self._coeffs[i]
+                else:
+                    non_diagonal_obs.append(t)
+                    non_diagonal_coeffs.append(self._coeffs[i])
```

Comment:
This sounds like it could be simplified by using a dictionary for `non_diagonal_obs`/`non_diagonal_coeffs`, since the order doesn't matter :)